### ISSUE TITLE
huge performance differences between gbm.train / gbm.predict vs LGBMClassifier fit / predict_proba w/ same hyper-parameters

### ISSUE THREAD

**Issue by flamby** [microsoft/LightGBM#2930]
Hi,

On a dataset I can't share unfortunately, I see huge differences in term of binary classification precision / recall scores between lightgbm native python api vs its sklearn wrapper.
The sklearn wrapper leads in my case to much better performance in term of precision / recall.

Another strange behaviour, is that the very same prediction (same random_state seed) invoked on machines with different CPU generations leads to huge different results (tested w/ the sklearn wrapper only).
For instance, the problem seems to have been worsen starting from lightgbm==2.1.2  on old architectures, whereas on new cpu architectures, starting from 2.1.2, performance improved. Any thought of major changes in 2.1.2 than could lead to huge performance differences on different cpu generations using pre-built wheel packages? [Issue 1487 suggesting to pass `boost_from_average=False`](https://github.com/Microsoft/LightGBM/issues/1487)did not work.

I'm certainly missing something obvious to you guys.

I tested w/ several lightgbm versions (2.3.1, 2.3.0, 2.2.3, latest master branch as of yesterday), w/ sklearn==0.21.3 which is the max version mandatory for lightgbm future 2.3.2)

Here is what I can share :

**Hyper-parameters used by both implementation**
```python
model_params = {'boosting_type': 'gbdt', 'colsample_bytree': 0.9500000000000001, 'learning_rate': 0.1,
                'max_depth': -1, 'min_child_samples': 89, 'min_child_weight': 0.001, 'min_split_gain': 0.0,
                'n_jobs': -1, 'num_leaves': 127, 'objective': 'binary', 'random_state': 0, 'seed': 0,
                'reg_alpha': 0.030559559479313415,
                'reg_lambda': 1.0949294490500017e-06, 'subsample': 0.8, 'subsample_for_bin': 200000,
                'subsample_freq': 4, 'verbose': -1, 'metric': 'binary_logloss'}
```

**Native python implementation**
```python
lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
gbm = lgb.train(model_params,
                            lgb_train,
                            num_boost_round=20,
                            valid_sets=lgb_eval,
                            early_stopping_rounds=5)
probas = gbm.predict(X_test)
# then i'm using these probabilities to find the best precision / recall tradeoff using the very same
# algorithm between the native api method and the sklearn wrapper method
```

**sklearn api implementation**
```python
model = lgb.LGBMClassifier(**model_params)
model.fit(X_train, y_train)
probas = model.predict_proba(X_test)
# then i'm using these probabilities to find the best precision / recall tradeoff using the very same
# algorithm between the native api method and the sklearn wrapper method
```

I see that in the sklearn implementation, I don't have the `num_boost_round` and `early_stopping_rounds` params.

Is there any parameters' names difference between the two, as it is the case in xgboost?

Thanks, and keep the good work!

**Comment by guolinke** [comment]
for the fair comparison, the sklearn version should be.
```python
model = lgb.LGBMClassifier(n_estimators=20, **model_params)
model.fit(X_train, y_train, eval_set=[X_test, y_test], early_stopping_rounds=5)
probas = model.predict_proba(X_test)
```

**Comment by flamby** [comment]
Thanks @guolinke 

Indeed, it reduced the gap a bit, but I still don't have the same results.
I just spotted that native api seeding is `seed` instead of `random_state`, so I fixed it too.

> for the fair comparison, the sklearn version should be.
> 
> ```python
> model = lgb.LGBMClassifier(n_estimators=20, **model_params)
> model.fit(X_train, y_train, eval_set=[(X_test, y_test)], early_stopping_rounds=5)
> probas = model.predict_proba(X_test)
> ```

**Comment by raviprasad1210** [comment]
Observed the same. The results are not matching between sklearn version and Python API.
What could be the possible explanation for this?

**Comment by guolinke** [comment]
@raviprasad1210 please ensure the hypermeters are the same. especially for sklearn interface, some of hyperparameters are in Constructor, some of them are in fit.

**Comment by raviprasad1210** [comment]
Hi. I am not getting same scores for lgb.train() and lgb.cv(). 
The problem is related to binning 
refer this: https://stackoverflow.com/questions/54709800/why-cant-i-match-lgbms-cv-score

The issue with this is - we generally use lgb.cv with full dataset as reference for estimating number of rounds and then tune remaining parameters using sklearn lgbmclassifier. If both numbers not matching then how do we proceed with hyper-parameter tuning

**Comment by StrikerRUS** [comment]
Our CI tests ensure that for the same params we have exact the same results on one machine for both standard and scikit-learn APIs. So I believe that the problem is somewhere in a way parameters are provided for different APIs in a code.
Unfortunately, without a MCVE it's impossible to provide further help.

https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_consistency.py
https://github.com/microsoft/LightGBM/blob/02a0089fc1c627e9f98776ab64c2aaac5209a56e/tests/python_package_test/test_sklearn.py#L393
